### PR TITLE
Remove read-only files under TempDir

### PR DESCRIPTION
**Pull request by vlsi**
Removal of read-only files fails (e.g. for Windows),
so the engine should try making the file writable first.

see #2046

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [ ] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

**Comment by vlsi** [comment]
Ok, cool. It reproduced with CI / Windows: https://github.com/junit-team/junit5/runs/250406301#step:4:746

**Comment by vlsi** [comment]
This looks good to me. @marcphilipp , would you please review it?

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TempDirectory.java [hunk 1]
```diff
@@ -190,8 +190,41 @@ public FileVisitResult postVisitDirectory(Path dir, IOException exc) {
...
+					}
+					catch (Exception e) {
+						// In case setWritable throws we can't proceed
+						ex.addSuppressed(e);
+						failures.put(path, ex);
+						return CONTINUE;
+					}
+					if (!isWritable) {
+						ex.addSuppressed(
+							new Exception("Cannot make file " + path + " writable with .toFile().setWritable(true)") {
```
Comment (marcphilipp) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TempDirectory.java]:
```suggestion
							new Exception("Attempt to make file '" + path + "' writable failed") {
```

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TempDirectory.java [hunk 2]
```diff
@@ -190,8 +190,41 @@ public FileVisitResult postVisitDirectory(Path dir, IOException exc) {
 				private FileVisitResult deleteAndContinue(Path path) {
 					try {
 						Files.delete(path);
+						return CONTINUE;
 					}
 					catch (IOException ex) {
+						return deleteUndeletableAndContinue(path, ex);
+					}
```
Comment (marcphilipp) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TempDirectory.java]: Can we move `return CONTINUE;` to the end of this method and make `deleteUndeletableAndContinue` return `void`?

Comment (vlsi) [reply junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TempDirectory.java]: I've made it return Throwable, so it is always recorded in a single place

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TempDirectory.java [hunk 3]
```diff
@@ -192,10 +192,44 @@ private FileVisitResult deleteAndContinue(Path path) {
...
+					catch (Exception e) {
+						// In case setWritable throws we can't proceed
+						ex.addSuppressed(e);
+						return ex;
+					}
+					if (!isWritable) {
+						ex.addSuppressed(new Exception("Attempt to make file '" + path + "' writable failed") {
+							@Override
+							public synchronized Throwable fillInStackTrace() {
+								return this; // Make the output smaller by omitting the stacktrace
```
Comment (sormuras) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TempDirectory.java]: Never saw this "blank stack trace" construct before.

Move the local class to a named class and extend ... `UnsupportedOperationException`? 🤔

**Comment by marcphilipp** [comment]
I refactored this a bit, I hope you don't mind, @vlsi.

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TempDirectory.java (hunk 3 above)
Comment (marcphilipp) [reply junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TempDirectory.java]: `UnwritableFileException`?

Comment (marcphilipp) [reply junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TempDirectory.java]: I went with `UndeletableFileException`

**Comment by sormuras** [comment]
Thanks @vlsi -- I'll add a line to the release notes directly on `master`.

**Pull request merged**